### PR TITLE
Feature/dgi 105

### PR DESCRIPTION
**Pull request by MichaelJeppesen**
Added functionality to attach a file to a statement and download it again

### CODE REVIEW

**Review by CSCITM — CHANGES_REQUESTED** [review]
Generally we wont review before all checks are successful, please keep an eye on failing checks.
image

File: frontend/src/components/Accounts/AccountList/AccountListItemButtons/EditOptions/DownloadStatementFileBtn.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+const DownloadStatementFileBtn: FC<Props> = ({ statement }) => {
+  const { t } = useLocales();
+
+  const download = useCallback(async () => {
+    const statementClient = await genStatementClient();
+    const data = await statementClient.getStatementFile(statement.id);
+
+    downloadFile(new Blob([base64ToArrayBuffer(data.data)]), data.fileName);
+  }, []);
```
Comment (objmj-itminds) [inline frontend/src/components/Accounts/AccountList/AccountListItemButtons/EditOptions/DownloadStatementFileBtn.tsx:31]: I am pretty sure we already have a util for this. If not create one.

File: frontend/src/components/Accounts/AccountList/AccountListItemButtons/EditOptions/DownloadStatementFileBtn.tsx [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+  return (
+    <>
+      {statement.statementFileName && (
+        <MenuItem onClick={download}>{t("statementFile.download")}</MenuItem>
+      )}
```
Comment (objmj-itminds) [inline frontend/src/components/Accounts/AccountList/AccountListItemButtons/EditOptions/DownloadStatementFileBtn.tsx:37]: The logic for this should probably be outside this component. Instead, consider throwing an error before the return if there isn't a `statementFileName`.

File: frontend/src/components/Accounts/AccountList/AccountListItemButtons/EditOptions/UploadStatementFileModal.tsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  const [isProcessing, setIsProcessing] = useState(false);
+  const { fetchData } = useContext(ClientsContext);
+
+  const { t } = useLocales();
+
+  const toast = useToast();
+
+  const uploadFile = useCallback(async () => {
+    setIsProcessing(true);
+    console.log(file.name);
```
Comment (objmj-itminds) [inline frontend/src/components/Accounts/AccountList/AccountListItemButtons/EditOptions/UploadStatementFileModal.tsx:39]: Remember not to use `console.log` instead, try the other functions for more precise logging. Or simply remove if not needed during production runtime.

**Review by objmj-itminds — CHANGES_REQUESTED** [review]

File: backend/Web/Controllers/StatementController.cs [hunk 4]
```diff
@@ -134,5 +136,24 @@ public async Task<ConsentFileDto> GetConsentFile([FromQuery] int statementId)
...
+    {
+      await Mediator.Send(new UploadStatementFileCommand{
+        StatementId = id,
+        StatementFile = file
+      });
+      return NoContent();
+    }
+
+    [HttpGet("statement/{id}/file")]
+    public async Task<ActionResult<StatementFileDto>> GetStatementFile([FromRoute] int id)
```
Comment (CSCITM) [inline backend/Web/Controllers/StatementController.cs:152]: Look into file results
https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.fileresult?view=aspnetcore-6.0

**Review by CSCITM — CHANGES_REQUESTED** [review]

File: backend/Application/Statements/Queries/GetStatementFile/GetStatementFileQuery.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+        if (file == null)
+        {
+          throw new NotFoundException("Statement file for statement with id " + statement.Id + " was not found.");
+        }
+
+        byte[] data;
+        using (Stream fileStream = new FileStream(file, FileMode.Open))
+        {
+          data = new byte[fileStream.Length];
```
Comment (CSCITM) [inline backend/Application/Statements/Queries/GetStatementFile/GetStatementFileQuery.cs:60]: Have you considered using FileStreamResult in the controller to be able to pass the stream directly?

**Review by CSCITM — CHANGES_REQUESTED** [review]

**Comment by CSCITM** [comment]
image
Please check for merge conflicts and failed checks, it's happens in too many of your PR's.

**Pull request merged**